kolibri/plugins/coach/assets/src/composables/useExerciseResources.js

```diff
@@ -0,0 +1,213 @@
+ * will reflect that topic's descendants automatically (with a loading state set accordingly).
+ *
+ * This is intended for use in a use case where the user is navigating the content tree in order
+ * to select contents.
+ *
+ * This module also reflects the pattern used by the ContentNodeResource.fetchTree method which
+ * provides metadata with which we can easily implement lazy loading of content nodes. In short, we
+ * get a `more` object which gives us the paramaters we need to use to fetch the next batch of nodes
+ * if they're available under the current topic.
+ *
+ */
+export default function useExerciseResources() {
+  /** @type {ref<string>} The topicId for which we will reflect the child nodes */
+  const _currentTopicId = ref(null);
+
+  /** @type {ref<Boolean>} Whether we are currently fetching/processing the child nodes */
+  const _loading = ref(false);
+
+  const _kindsToFetch = ref([ContentNodeKinds.EXERCISE, ContentNodeKinds.TOPIC]);
+
+  /** @type {ref<Array>} The list of child nodes for the current topic. */
+  const _contentList = ref([]);
+
+  /** @type {ref<Object>} The params object we pass to the ContentNodeResource fetchTree method to
+   * fetch the next batch of nodes.
+   */
+  const _moreParams = ref(null);
+
+  const channels = ref([]);
+
+  const bookmarks = ref([]);
+
+  function _numAssessmentsInFetchedChildren(topic) {
+    if (!topic.children) {
+      return 0;
+    } else {
+      return topic.children.reduce((acc, child) => {
+        acc += child.children.length || 0;
+        return acc;
+      }, 0);
+    }
+  }
+
+  /**
+   * @param {Object} topic The topic object to check -- must be annotated with the `num_assessments`
+   * property -- which means you should be passing a topic that you found in the contentList
+   */
+  function shouldTopicHaveCheckbox(topic) {
+    const num_assessments = topic.num_assessments;
+    if (num_assessments === undefined) {
+      console.error(
+        'You cannot pass a topic that has not been annotated with the `num_assessments` property.',
+        'You should be passing a topic that you found in the contentList.'
+      );
+    }
+    if (get(_contentList).children) {
+      const childTopicIds = get(_contentList)
+        .filter(({ kind }) => kind === ContentNodeKinds.TOPIC)
+        .map(({ id }) => id);
+    }
+  }
+
+  function topicInChildrenHasMore() {}
+
+  function fetchMore() {
+    if (get(_moreParams)) {
+      fetchContentList(_moreParams);
+    } else {
+      console.error("No more params to fetch. There shouldn't be a 'Show More' button visible.");
+    }
+  }
+
+  /**
+   * @param {Boolean} appendResults Whether to append the results to the current list or replace it
+   */
+  function fetchContentList(moreParams = null) {
+    set(_loading, true);
+
+    // Set the default params for the fetchTree method .
+    const params = {
+      ...(moreParams || {}),
```

Comment:
Not that the `more` object from a previous paginated fetch will include the previously used query parameters, so adding both is slightly redundant.

It might be good to separate out the middle part of this function with the behaviour that is common to 'base' fetching and 'more' fetching into a helper function, and then create `fetchContentList` and `fetchMoreContent` functions that do the specific parameter injection, and result processing.